docs/build/use-cases/key-generation.mdx

```diff
@@ -0,0 +1,576 @@
+main().catch((e) => { console.error(e); process.exit(1); });
+```
+
+</details>
+
+## Hardhat (contracts only)
+
+Minimal config to compile `Counter.sol`:
+
+<details>
+  <summary>hardhat.config.cjs</summary>
+
+```cjs
+/** @type import('hardhat/config').HardhatUserConfig */
+module.exports = {
+  solidity: { version: "0.8.24", settings: { optimizer: { enabled: true, runs: 200 } } },
+  paths: { sources: "./contracts", artifacts: "./artifacts", cache: "./cache" }
+};
+```
+
+</details>
+
+Compile locally (optional):
+
+```shell
+npx hardhat compile
+```
+
+## Containerize
+
+Add a compose file that mounts the **appd socket** provided by ROFL and
+exposes port **8080**. We set Base Sepolia RPC defaults.
+
+<details>
+  <summary>compose.yaml</summary>
+
+```yaml
+services:
+  demo:
+    image: docker.io/YOURUSER/rofl-keygen:0.1.0
+    platform: linux/amd64
+    environment:
+      - PORT=${PORT:-8080}
+      - KEY_ID=${KEY_ID:-evm:base:sepolia}
+      - BASE_RPC_URL=${BASE_RPC_URL:-https://sepolia.base.org}
+      - BASE_CHAIN_ID=${BASE_CHAIN_ID:-84532}
+    ports:
+      - "8080:8080"
+    volumes:
+      - /run/rofl-appd.sock:/run/rofl-appd.sock
+```
+
+</details>
+
+Add a Dockerfile that builds TS and compiles the contract (optional):
+
+<details>
+  <summary>Dockerfile</summary>
+
+```dockerfile
+FROM node:20-alpine
+WORKDIR /app
+
+COPY package.json package-lock.json* ./
+RUN npm ci
+
+COPY tsconfig.json ./
+COPY src ./src
+COPY contracts ./contracts
+COPY hardhat.config.cjs ./
+RUN npm run build || npx tsc \
+ && npx hardhat compile \
+ && npm prune --omit=dev
+
+ENV NODE_ENV=production
+ENV PORT=8080
+EXPOSE 8080
+CMD ["node", "dist/server.js"]
+```
+
+</details>
+
+## Scripts
+
+Add handy scripts to `package.json`:
+
+```json
+{
+  "scripts": {
+    "build": "tsc -p tsconfig.json",
+    "start": "node dist/server.js",
+    "dev": "tsx src/server.ts",
+    "deploy-counter": "node scripts/deploy-contract.js"
+  }
+}
+```
+
+## Build the image
+
+ROFL runs on **x86_64/TDX**, so build amd64 images:
+
+```shell
+docker buildx build --platform linux/amd64 \
+  -t docker.io/YOURUSER/rofl-keygen:0.1.0 --push .
+```
+
+(Optionally pin the digest and use `image: ...@sha256:...` in compose.)
+
+## Build ROFL bundle
+
+<Tabs>
+  <TabItem value="Native Linux">
+    ```shell
+    oasis rofl build
+    ```
+  </TabItem>
+  <TabItem value="Docker (Mac/Windows/Linux)">
+    ```shell
+    docker run --platform linux/amd64 --volume .:/src \
+      -it ghcr.io/oasisprotocol/rofl-dev:main oasis rofl build
+    ```
+  </TabItem>
+</Tabs>
+
+Then publish the enclave identities and config:
+
+```shell
+oasis rofl update
+```
+
+## Deploy
+
+Deploy to a Testnet provider:
+
+```shell
+oasis rofl deploy
+```
+
+Find your public HTTPS URL:
+
+```shell
+oasis rofl machine show
+```
+
+Look for the **Proxy** section (e.g., `https://p8080.mXXX.test-proxy...`).
+
+## End‑to‑end (Base Sepolia)
+
+1. **Get App ID**
+
+   ```shell
+   curl -s https://YOUR-PROXY/app-id | jq
```

Comment:
`YOUR_PROXY` is the machine name defined in compose.yaml. Also, https won't work out of the box since the domain is not managed by CA and self-signed certs are not allowed.